### DISCUSSION TITLE
Second derivative as a boundary condition

### DISCUSSION THREAD

**Discussion by JarosT55** [kinnala/scikit-fem#1095] [Q&A] [answered]
Hi,

Is there 'a simple' way of defining ∇²=0 as a boundary condition? One option I see, is establishing pair (or triple) boundaries with ∇=0.
It there a better way ....?

Regards
MichaelT

**Comment by kinnala** [comment]
I think it is not mathematically valid boundary condition for second order PDE. Say you are solving something like $$\nabla^2 u = f.$$ Then it would just correspond to having $$f=0$$ somewhere but it doesn't help much if done at a single point or a line. You can set $f=0$ over a set of elements.

Fourth order PDE support something like this. What kind of PDE are you solving?

**Reply by JarosT55** [reply]
$\nabla^2 u = 0$       **only** on      $\Gamma$

Traditionally and by **skfem** examples **Navier-Stoke** equations are evaluated on **{'u': ElementVector(ElementTriP2()),'p': ElementTriP1()}**, 
thus, the second derivative evaluation of **u** will give a constant.
However, using higher degree **Tri's** should alleviate the problem.
Theoretically, implementing high-order boundary conditions should be possible with a family of high-order **ElementTriPX()**. Should it not?
How complex would such implementation be in  **skfem** ... like ☕?

**Reply by gdmcbain** [reply]
No, the higher order Lagrangian elements don't help. They are sufficiently differentiable in their interior but those derivatives aren't continuous between elements. What's needed is Hermite elements.

However, I echo the concern that this is not a mathematically or physically valid boundary condition for the Navier–Stokes or Stokes equation. When you say 'nonlinear' Navier-Stokes what do you mean? Not merely the usual quasi linearity of the quadratic advection term but a nonnewtonian viscous term? The valid boundary conditions all pretty much arise from the integration by parts of the viscous term.

**Reply by JarosT55** [reply]
**Hi,**

**It is how I understand the issue.**
By simple integration, we got.      $({\Omega-i})$  denotes a 'slice' of ${\Omega}$ space

$\nabla^2 u = f(r)$ 

$\nabla u = \int_{\Omega-1} (\nabla^2 u) \mathrm{d}V_{-1} + C$ 

$u = \int_{\Omega-2} (\nabla u) \mathrm{d}V_{-2} = \int_\Omega (\nabla^2 u) \mathrm{d}V + \int_{\Omega-2} C \mathrm{d}V_{-2} + D = \int_\Omega f(r) \mathrm{d}V + \int_{\Omega-2} C \mathrm{d}V_{-2} + D $


This leads to $u = 0$  for boundary conditions $\nabla^2 u = 0$,  when $C$ and $D$ integration constants are equal $0$. 

However, they don't need to be $0$, or $0$ on the whole domain. Besides, there might be other equation's constraints also.
I am trying **'intuitively'** to imagine the case of such function $f(r)$, which has a region where $\nabla^2 u = 0$ on some $\subset\Omega$ and $\neq 0$ outside. 
It should topologically fall into categories of saddle functions, periodic ones, etc.
Does the solution for an equally charged dipole represent such a function?

**Regards
MichaelT**

**Reply by JarosT55** [reply]
**Hi Mr. GDMcBain,**

_"No, the higher order Lagrangian elements don't help. They are sufficiently differentiable in their interior but those derivatives aren't continuous between elements. What's needed is Hermite elements."_

I found a very interesting elaboration about local vs. global continuity of in-element-functions and  their globally mapped counterparts :
https://math.stackexchange.com/questions/2402843/if-the-symbol-partial-omega-is-used-to-represent-the-boundary-of-omega-su

I am still digesting the content of it, but as I currently understand, TriP4 should offer globally continuous mapping of the second derivative.
I will extend my thought processes ... while in a sleep 💤!

**Regards
MichaelT**

**Reply by gdmcbain** [reply]
The difference between Hermite and Lagrange elements is that the degrees of freedom of the latter are all ordinates (at boundary or internal points) while those of the former include the values of derivatives at the boundary; it's this last property that's required for continuity between elements.

**Reply by JarosT55** [reply]
**Hi Mr. GDMcBain,**

It is a valid point…triggering me to find an analogy to better understand **Lagrange** and **Hermite** element types.
I visualize a balloon with a mesh wrapped around it. In the case of Lagrangians, the mesh approximates a shape along the mesh's edges to the derivative degree level dependent upon the order of **ElementTriX()** in our case. What is happening inside meshes eyes … is the _"Alice in Wonderland"_ case … but should we care? … later on this subject. What is important is that we can reliably integrate along mesh edges and construct a smooth enough function representing a path on the balloon's surface, even if the respective element's mapped globally function protrudes through the mesh.

**Hermite** elements with additional constrainement of their quadrature points (as I understand) can estimate the surface function (and its derivatives) inside mesh eyes.
It means that we will be able to integrate reliably along the element perimeter but also find out the surface area of the balloon patch defined by the element. I guess **Hermites** are more expensive than **Lagrangians**.

What does the above (if my understanding is correct) mean in the **basic Navier-Stoke PDE** context?
What are we looking for there?
My perception is that we try to solve equalities:
•	Local pressure as a drop of stream velocity - $\nabla ≃ p$
•	Local pressure divergence corresponding vector Laplacian of u - $\nabla^2 ≃ \nabla p$
As I understand, _**we search for such equalities along vectors defined by mesh edges**_. What happens inside the element is primarily irrelevant… unless we want to find, for example, the forces associated along the hypothetical stream … or try 'detailed' $p$ by nipping inside the cell by selecting a higher order of element $p$ like, e.g. **ElementTriP2()** (in combination with $u:$ **ElementVector(ElementTriP4())** ); the second PNG (pressure_A_4_2.png). Trend of the function is almost the same, but **'protrusions' through a mesh are very visible**!.

I assume that the choice of **ElementTri()** types for the classical **Navier-Stoke PDE** is the consequence of the above … and concerns about algorithm speed and stability. 
![pressure_A_1_2](https://github.com/kinnala/scikit-fem/assets/23690439/351e0bf4-10c0-4cd3-9be7-c5f893fdd083)
![pressure_A_4_2](https://github.com/kinnala/scikit-fem/assets/23690439/ff04b307-0c89-437c-8c16-a5a51d70adf7)
**Regards
MichaelT**

**Comment by JarosT55** [comment]
Nonlinear Navier-Stokes.

**Reply by JarosT55** [reply]
**Hi Mr Kinnala,** 

**Idea** ...
_scipy.linalg.solve()_ offers _assume_a_ parameter facilitating dealing with symmetric matrices.
Is there a way to trick or build in a **skfem** environment creating a similar mechanism for _scipy.sparse.linalg.spsolve()_?

**Skfem** offers predefined meshes with symmetries. Can & how this could be extended to custom meshes?

**Regards
MichaelT**

**Reply by gdmcbain** [reply, marked as answer]
The answer is yes but better to start a new discussion for this new question.

**Reply by JarosT55** [reply]
Yes, Mr. GDMcBain, I will ⌛!
